File: android/app/src/main/kotlin/com/edgelink/core/PairingStore.kt
package com.edgelink.core

import java.time.Instant

data class PinnedPeer(
    val deviceId: String,
    val name: String,
    val publicKey: ByteArray,
    val pairedAt: Instant
) {
    init {
        require(DeviceId.isValid(deviceId)) { "Device ID must be 9 digits without a leading zero." }
        require(publicKey.size == 32) { "Ed25519 public key must be 32 bytes." }
    }
}

interface PairingStore {
    fun loadPeer(deviceId: String): PinnedPeer?
    fun savePeer(peer: PinnedPeer)
}

class InMemoryPairingStore : PairingStore {
    private val peers = mutableMapOf<String, PinnedPeer>()

    override fun loadPeer(deviceId: String): PinnedPeer? = peers[deviceId]

    override fun savePeer(peer: PinnedPeer) {
        peers[peer.deviceId] = peer
    }
}

File: mac/Sources/EdgeLinkKit/PairingStore.swift
import Foundation

public struct PinnedPeer: Codable, Equatable, Sendable {
    public let deviceId: String
    public let name: String
    public let publicKey: Data
    public let pairedAt: Date

    public init(deviceId: String, name: String, publicKey: Data, pairedAt: Date) {
        precondition(DeviceID.isValid(deviceId), "Device ID must be 9 digits without a leading zero.")
        self.deviceId = deviceId
        self.name = name
        self.publicKey = publicKey
        self.pairedAt = pairedAt
    }
}

public protocol PairingStore: Sendable {
    func loadPeer(deviceId: String) throws -> PinnedPeer?
    func savePeer(_ peer: PinnedPeer) throws
}

public final class InMemoryPairingStore: PairingStore, @unchecked Sendable {
    private var peers: [String: PinnedPeer] = [:]

    public init() {}

    public func loadPeer(deviceId: String) throws -> PinnedPeer? {
        peers[deviceId]
    }

    public func savePeer(_ peer: PinnedPeer) throws {
        peers[peer.deviceId] = peer
    }
}
